examples/tic-tac-toe/ui/src/components/NewSharedGame.tsx

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useCurrentAccount } from '@mysten/dapp-kit';
+import { isValidSuiAddress, normalizeSuiAddress } from '@mysten/sui/utils';
+import { ExclamationTriangleIcon } from '@radix-ui/react-icons';
+import { Box, Button, Em, Flex, Separator, Text, TextField } from '@radix-ui/themes';
+import { useExecutor } from 'hooks/useExecutor';
+import { useTransactions } from 'hooks/useTransactions';
+import { ReactElement, useState } from 'react';
+
+import { ComputedField } from './ComputedField';
+
+/**
+ * Form for creating a new shared game.
+ */
+export function NewSharedGame() {
+	// SAFETY: <App /> tests that a package exists, so Transactions
+	// builder should be available.
+	const tx = useTransactions()!!;
+	const signAndExecute = useExecutor();
+
+	const player = useCurrentAccount()?.address;
+	const [opponent, setOpponent] = useState<string | null>(null);
+
+	const hasPlayer = player != null;
+	const hasOpponent = opponent != null;
+
+	function onClick() {
+		signAndExecute(
+			{
+				// SAFETY: Button is only enabled when player and opponent are
+				// available.
+				tx: tx.newSharedGame(player!!, opponent!!),
+				options: { showEffects: true },
+			},
+			({ effects }) => {
+				const gameId = effects?.created?.[0].reference?.objectId;
+				if (gameId) {
+					window.location.href = `/${gameId}`;
+				}
+			},
+		);
+	}
+
+	return (
+		<>
+			<ComputedField label="Your address" value={player} />
+			<TextField.Root
+				size="2"
+				mb="2"
+				placeholder="Opponent address"
+				style={{ width: '100%' }}
+				color={hasOpponent ? undefined : 'red'}
+				variant={hasOpponent ? 'surface' : 'soft'}
+				onChange={(e) => setOpponent(normalizedAddress(e.target.value))}
+			/>
+			<Flex justify="between" mt="4">
+				<Validation hasPlayer={hasPlayer} hasOpponent={hasOpponent} />
+				<Button variant="outline" disabled={!(player && opponent)} onClick={onClick}>
+					Play
+				</Button>
+			</Flex>
+			<Separator orientation="horizontal" my="4" style={{ width: '100%' }} />
+			<Em>
+				<Text as="div">
+					Create the game as a shared object that both players can access. Each move is a single
+					transaction, but it requires going through consensus.
+				</Text>
+			</Em>
+		</>
+	);
+}
+
+function Validation({
+	hasPlayer,
+	hasOpponent,
+}: {
+	hasPlayer: boolean;
+	hasOpponent: boolean;
+}): ReactElement {
+	if (!hasPlayer) {
+		return (
+			<Flex align="center" gap="2">
+				<ExclamationTriangleIcon color="red" />
+				<Text color="red">Wallet not connected.</Text>
+			</Flex>
+		);
+	} else if (!hasOpponent) {
+		return (
+			<Flex align="center" gap="2">
+				<ExclamationTriangleIcon color="red" />
+				<Text color="red">Invalid opponent address.</Text>
+			</Flex>
+		);
+	} else {
+		return <Box />;
+	}
```

Comment:
super nit: can skip the "else" statement and just return <Box />.

Theoretically they can all be top-level ifs without `if,elseif,else` but that might be my personal preference.

Reply:
The `<Box />` is necessary to make sure there's something on the left hand side of the flex box -- otherwise the "Play" button this is sitting next to will get pulled to the left when there are no validation issues -- but I'll switch this to an early return style!

Reply:
Oh that's exactly what I meant (early return versus nested if-elseif-else statements). But that's just me not liking nesting 😄

Reply:
Ah, got it!

Reply:
I thought you were saying just `return;` (i.e. don't need to return anything instead of `Box`), but I understand now!